File: back-end/src/main/java/com/example/backend/service/impl/VehicleServiceImpl.java
package com.example.backend.service.impl;

import com.example.backend.entities.School;
import com.example.backend.entities.Vehicle;
import com.example.backend.exception.EntityNotFoundException;
import com.example.backend.exception.ValidationException;
import com.example.backend.repository.SchoolRepository;
import com.example.backend.repository.VehicleRepository;
import com.example.backend.service.VehicleService;
import com.example.backend.service.base.BaseServiceImpl;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@Transactional(readOnly = true)
public class VehicleServiceImpl extends BaseServiceImpl<Vehicle, VehicleRepository> implements VehicleService {

    private final SchoolRepository schoolRepository;

    public VehicleServiceImpl(VehicleRepository repository, SchoolRepository schoolRepository) {
        super(repository);
        this.schoolRepository = schoolRepository;
    }

    @Override
    public Optional<Vehicle> findByLicensePlate(String licensePlate) {
        log.debug("Finding vehicle by license plate: {}", licensePlate);
        return repository.findByLicensePlateAndActiveTrue(licensePlate);
    }

    @Override
    public List<Vehicle> findByAvailable(boolean available) {
        log.debug("Finding vehicles by availability: {}", available);
        return repository.findAvailableAndActiveVehicles();
    }

    @Override
    public Optional<Vehicle> findByIdWithDriver(Long vehicleId) {
        log.debug("Finding vehicle {} with driver", vehicleId);
        return repository.findByIdWithDriver(vehicleId);
    }

    @Override
    public List<Vehicle> findAvailableVehicles() {
        log.debug("Finding all available vehicles");
        return repository.findAvailableVehicles();
    }

    @Override
    public List<Vehicle> findVehiclesBySchool(Long schoolId) {
        log.debug("Finding vehicles for school {}", schoolId);
        validateSchoolExists(schoolId);
        return repository.findVehiclesBySchool(schoolId);
    }

    @Override
    public List<Vehicle> findNearbyVehicles(Double latitude, Double longitude, Double radiusInMeters) {
        log.debug("Finding vehicles near ({}, {}) within {} meters", latitude, longitude, radiusInMeters);
        validateCoordinates(latitude, longitude);
        if (radiusInMeters == null || radiusInMeters <= 0) {
            throw new ValidationException("radius", "Radius must be greater than 0");
        }
        return repository.findNearbyVehicles(latitude, longitude, radiusInMeters);
    }

    @Override
    public List<Vehicle> findActivelyOperatingVehicles() {
        log.debug("Finding actively operating vehicles");
        return repository.findActivelyOperatingVehicles();
    }

    @Override
    public long countCompletedTrips(Long vehicleId) {
        log.debug("Counting completed trips for vehicle {}", vehicleId);
        validateVehicleExists(vehicleId);
        return repository.countCompletedTrips(vehicleId);
    }

    @Override
    @Transactional
    public void updateLocation(Long vehicleId, Double latitude, Double longitude) {
        log.debug("Updating location for vehicle {} to ({}, {})", vehicleId, latitude, longitude);
        Vehicle vehicle = findVehicleById(vehicleId);
        validateCoordinates(latitude, longitude);

        vehicle.setCurrentLatitude(latitude);
        vehicle.setCurrentLongitude(longitude);

        repository.save(vehicle);
        log.info("Successfully updated location for vehicle {}", vehicleId);
    }

    @Override
    @Transactional
    public void markAsAvailable(Long vehicleId) {
        log.debug("Marking vehicle {} as available", vehicleId);
        Vehicle vehicle = findVehicleById(vehicleId);

        if (vehicle.getDriver() != null) {
            throw new ValidationException("Cannot mark vehicle as available while assigned to a driver");
        }
        if (vehicle.getStatus() == Vehicle.VehicleStatus.MAINTENANCE) {
            throw new ValidationException("Cannot mark vehicle as available while under maintenance");
        }

        vehicle.setStatus(Vehicle.VehicleStatus.AVAILABLE);
        repository.save(vehicle);
        log.info("Successfully marked vehicle {} as available", vehicleId);
    }

    @Override
    @Transactional
    public void markAsUnavailable(Long vehicleId) {
        log.debug("Marking vehicle {} as unavailable", vehicleId);
        Vehicle vehicle = findVehicleById(vehicleId);

        if (vehicle.getStatus() == Vehicle.VehicleStatus.IN_USE) {
            throw new ValidationException("Cannot mark vehicle as unavailable while in use");
        }
        if (vehicle.getStatus() == Vehicle.VehicleStatus.MAINTENANCE) {
            throw new ValidationException("Cannot mark vehicle as unavailable while under maintenance");
        }

        vehicle.setStatus(Vehicle.VehicleStatus.OUT_OF_SERVICE);
        repository.save(vehicle);
        log.info("Successfully marked vehicle {} as unavailable", vehicleId);
    }

    @Override
    @Transactional
    public void assignToSchool(Long vehicleId, Long schoolId) {
        log.debug("Assigning vehicle {} to school {}", vehicleId, schoolId);
        Vehicle vehicle = findVehicleById(vehicleId);
        School school = findSchoolById(schoolId);

        if (vehicle.getSchool() != null) {
            if (vehicle.getSchool().getId().equals(schoolId)) {
                throw new ValidationException("Vehicle is already assigned to this school");
            }
            throw new ValidationException("Vehicle is already assigned to another school");
        }

        vehicle.setSchool(school);
        repository.save(vehicle);
        log.info("Successfully assigned vehicle {} to school {}", vehicleId, schoolId);
    }

    @Override
    @Transactional
    public void removeFromSchool(Long vehicleId) {
        log.debug("Removing vehicle {} from school assignment", vehicleId);
        Vehicle vehicle = findVehicleById(vehicleId);

        if (vehicle.getSchool() == null) {
            throw new ValidationException("Vehicle is not assigned to any school");
        }

        vehicle.setSchool(null);
        repository.save(vehicle);
        log.info("Successfully removed vehicle {} from school assignment", vehicleId);
    }

    @Override
    @Transactional
    public void performMaintenance(Long vehicleId, String maintenanceType, String notes) {
        log.debug("Starting maintenance for vehicle {}: {}", vehicleId, maintenanceType);
        Vehicle vehicle = findVehicleById(vehicleId);

        if (vehicle.getStatus() == Vehicle.VehicleStatus.IN_USE) {
            throw new ValidationException("Cannot start maintenance while vehicle is in use");
        }
        if (vehicle.getStatus() == Vehicle.VehicleStatus.MAINTENANCE) {
            throw new ValidationException("Vehicle is already under maintenance");
        }
        if (!StringUtils.hasText(maintenanceType)) {
            throw new ValidationException("maintenanceType", "Maintenance type is required");
        }

        vehicle.setStatus(Vehicle.VehicleStatus.MAINTENANCE);
        repository.save(vehicle);
        log.info("Successfully started maintenance for vehicle {}", vehicleId);
    }

    @Override
    @Transactional
    public void completeMaintenance(Long vehicleId, String notes) {
        log.debug("Completing maintenance for vehicle {}", vehicleId);
        Vehicle vehicle = findVehicleById(vehicleId);

        if (vehicle.getStatus() != Vehicle.VehicleStatus.MAINTENANCE) {
            throw new ValidationException("Vehicle is not under maintenance");
        }

        vehicle.setStatus(Vehicle.VehicleStatus.AVAILABLE);
        vehicle.setLastMaintenanceDate(LocalDateTime.now().toLocalDate());
        repository.save(vehicle);
        log.info("Successfully completed maintenance for vehicle {}", vehicleId);
    }

    @Override
    public List<Vehicle> findVehiclesWithStats() {
        log.debug("Finding vehicles with statistics");
        List<Vehicle> vehicles = repository.findAllByActiveTrue();
        vehicles.forEach(vehicle -> {
            vehicle.setCompletedTripsCount(countCompletedTrips(vehicle.getId()));
        });
        return vehicles;
    }

    @Override
    @Transactional
    public void delete(Long id) {
        log.debug("Deleting vehicle {}", id);
        Vehicle vehicle = findVehicleById(id);
        
        if (vehicle.getStatus() == Vehicle.VehicleStatus.IN_USE) {
            throw new ValidationException("Cannot delete vehicle while it is in use");
        }
        if (vehicle.getDriver() != null) {
            throw new ValidationException("Cannot delete vehicle while assigned to a driver");
        }

        repository.delete(vehicle);
        log.info("Successfully deleted vehicle {}", id);
    }

    // These methods are not applicable for vehicles
    @Override
    public Vehicle createDriver(Vehicle entity) {
        throw new UnsupportedOperationException("Operation not supported for vehicles");
    }

    @Override
    public void updateDriverProfile(Long id, Vehicle entity) {
        throw new UnsupportedOperationException("Operation not supported for vehicles");
    }

    // Private helper methods
    private void validateCoordinates(Double latitude, Double longitude) {
        if (latitude == null || longitude == null) {
            throw new ValidationException("Latitude and longitude are required");
        }
        if (latitude < -90 || latitude > 90) {
            throw new ValidationException("latitude", "Invalid latitude value");
        }
        if (longitude < -180 || longitude > 180) {
            throw new ValidationException("longitude", "Invalid longitude value");
        }
    }

    private void validateSchoolExists(Long schoolId) {
        if (!schoolRepository.existsById(schoolId)) {
            throw new EntityNotFoundException("School", schoolId);
        }
    }

    private void validateVehicleExists(Long vehicleId) {
        if (!repository.existsById(vehicleId)) {
            throw new EntityNotFoundException("Vehicle", vehicleId);
        }
    }

    private Vehicle findVehicleById(Long vehicleId) {
        return repository.findById(vehicleId)
                .orElseThrow(() -> new EntityNotFoundException("Vehicle", vehicleId));
    }

    private School findSchoolById(Long schoolId) {
        return schoolRepository.findById(schoolId)
                .orElseThrow(() -> new EntityNotFoundException("School", schoolId));
    }
} 

File: front-end/src/services/impl/VehicleServiceImpl.ts
import { VehicleService } from '../VehicleService';
import { Vehicle } from '../../core/entities/vehicle.entity';
import { CreateVehicleRequest, UpdateVehicleRequest } from '../../core/dto/vehicle.dto';
import { BaseHttpService } from '../BaseHttpService';
import { ApiResponse } from '../../core/models/ApiResponse';

export class VehicleServiceImpl extends BaseHttpService implements VehicleService {
  constructor() {
    super('/v1/vehicles');
  }

  async findAll(): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>();
    return response.data;
  }

  async findById(id: number): Promise<Vehicle> {
    const response = await this.get<ApiResponse<Vehicle>>(`/${id}`);
    return response.data;
  }

  async create(vehicleRequest: CreateVehicleRequest): Promise<Vehicle> {
    // Ensure coordinates are passed correctly to both fields
    if (vehicleRequest.initialLatitude !== undefined && vehicleRequest.initialLatitude !== 0) {
      vehicleRequest.currentLatitude = vehicleRequest.initialLatitude;
    }
    
    if (vehicleRequest.initialLongitude !== undefined && vehicleRequest.initialLongitude !== 0) {
      vehicleRequest.currentLongitude = vehicleRequest.initialLongitude;
    }
    
    const response = await this.post<ApiResponse<Vehicle>>('', vehicleRequest);
    return response.data;
  }

  async addVehicle(formData: any): Promise<Vehicle> {
    const vehicleRequest: CreateVehicleRequest = {
      vehicleNumber: formData.licensePlate,
      make: formData.make,
      model: formData.model,
      year: formData.year,
      capacity: formData.capacity,
      status: formData.status,
      currentMileage: formData.currentMileage,
      insuranceExpiryDate: formData.registrationExpiryDate,
      registrationExpiryDate: formData.registrationExpiryDate,
      lastMaintenanceDate: new Date().toISOString().split('T')[0],
      schoolId: formData.schoolId,
      initialLatitude: formData.currentLatitude,
      initialLongitude: formData.currentLongitude
    };
    
    return this.create(vehicleRequest);
  }

  async update(id: number, vehicleRequest: UpdateVehicleRequest): Promise<Vehicle> {
    // If there are latitude/longitude properties, map them to initialLatitude/initialLongitude as well
    if (vehicleRequest.currentLatitude !== undefined || vehicleRequest.currentLongitude !== undefined) {
      vehicleRequest = {
        ...vehicleRequest,
        initialLatitude: vehicleRequest.currentLatitude,
        initialLongitude: vehicleRequest.currentLongitude
      };
    }
    
    const response = await this.put<ApiResponse<Vehicle>>(`/${id}`, vehicleRequest);
    return response.data;
  }

  async deleteVehicle(id: number): Promise<void> {
    await this.delete<ApiResponse<void>>(`/${id}`);
  }

  async findByLicensePlate(licensePlate: string): Promise<Vehicle> {
    const response = await this.get<ApiResponse<Vehicle>>(`/license-plate/${encodeURIComponent(licensePlate)}`);
    return response.data;
  }

  async findAvailableVehicles(): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>('/available');
    return response.data;
  }

  async findVehiclesBySchool(schoolId: number): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>(`/school/${schoolId}`);
    return response.data;
  }

  async findNearbyVehicles(latitude: number, longitude: number, radiusInMeters: number): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>(
      `/nearby?latitude=${latitude}&longitude=${longitude}&radiusInMeters=${radiusInMeters}`
    );
    return response.data;
  }

  async findActivelyOperatingVehicles(): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>('/operating');
    return response.data;
  }

  async countCompletedTrips(id: number): Promise<number> {
    const response = await this.get<ApiResponse<number>>(`/${id}/trips/count`);
    return response.data;
  }

  async updateLocation(id: number, latitude: number, longitude: number): Promise<void> {
    await this.put<ApiResponse<void>>(`/${id}/location?latitude=${latitude}&longitude=${longitude}`);
  }

  async markAsAvailable(id: number): Promise<void> {
    await this.put<ApiResponse<void>>(`/${id}/available`);
  }

  async markAsUnavailable(id: number): Promise<void> {
    await this.put<ApiResponse<void>>(`/${id}/unavailable`);
  }

  async assignToSchool(id: number, schoolId: number): Promise<void> {
    await this.put<ApiResponse<void>>(`/${id}/assign-school/${schoolId}`);
  }

  async removeFromSchool(id: number): Promise<void> {
    await this.put<ApiResponse<void>>(`/${id}/remove-school`);
  }

  async performMaintenance(id: number, maintenanceType: string, notes: string): Promise<void> {
    await this.post<ApiResponse<void>>(
      `/${id}/maintenance?maintenanceType=${encodeURIComponent(maintenanceType)}&notes=${encodeURIComponent(notes)}`
    );
  }

  async completeMaintenance(id: number, notes: string): Promise<void> {
    await this.post<ApiResponse<void>>(`/${id}/maintenance/complete?notes=${encodeURIComponent(notes)}`);
  }

  async findVehiclesWithStats(): Promise<Vehicle[]> {
    const response = await this.get<ApiResponse<Vehicle[]>>('/with-stats');
    return response.data;
  }
}